File: blunder/src/robot/robot.go
package robot

import (
	"fmt"
	"os"
)

type Robot struct {
	posX                  int
	posY                  int
	invert                bool
	drunk                 bool
	currentDirection      string
	direction             [4]string
	currentDirectionIndex int
	potentialLoop         bool
	loop                  bool
	memory                []string
	movement              []string
}

func NewRobot(posX int, posY int) *Robot {
	b := new(Robot)
	b.posX = posX
	b.posY = posY
	b.invert = false
	b.drunk = false
	b.currentDirection = "S"
	b.direction = [4]string{"S", "E", "N", "W"}
	b.currentDirectionIndex = 0
	b.potentialLoop = false
	b.loop = false
	return b
}

func (r *Robot) Move(x int, y int) {
	r.posX = x
	r.posY = y
	r.movement = append(r.movement, r.GetCurrentDirectionString())
	r.MemorizeMovement()
}

func GetOppositeDirection(direction string) string {
	switch direction {
	case "S":
		return "N"
	case "E":
		return "W"
	case "N":
		return "S"
	case "W":
		return "E"
	}
	return "error"
}

// Memorize movement and return if the robot is looping
func (r *Robot) MemorizeMovement() {
	if len(r.memory) == 0 {
		r.memory = append(r.memory, r.currentDirection)
	} else {
		if r.memory[0] == r.currentDirection { // The robot continues his path
			r.memory = append(r.memory, r.currentDirection)
		} else if r.memory[0] == GetOppositeDirection(r.currentDirection) { // The robot comes back
			r.memory = append(r.memory, r.currentDirection)
			r.IsStuckInLoop()
		} else { // The robot can't be in loop in this case
			r.memory = nil
			r.memory = append(r.memory, r.currentDirection) // It starts his new journey
		}
	}
}

func (r *Robot) CleanInitialDirectionFromMemory(oppositeDirectionCount int, oppositeDirection string) {
	r.memory = nil
	for i := 0; i < oppositeDirectionCount; i++ {
		r.memory = append(r.memory, oppositeDirection)
	}
}

func (r *Robot) IsStuckInLoop() {
	initialDirection := r.memory[0]
	initialDirectionCount := 0
	oppositeDirection := GetOppositeDirection(initialDirection)
	oppositeDirectionCount := 0
	for _, val := range r.memory {
		if val == initialDirection {
			initialDirectionCount++
		} else if val == oppositeDirection {
			oppositeDirectionCount++
		} else {
			fmt.Fprintln(os.Stdout, "Memory is compromised...")
		}
	}
	if initialDirectionCount == oppositeDirectionCount {
		if r.potentialLoop {
			r.loop = true
		} else {
			r.potentialLoop = true
			r.CleanInitialDirectionFromMemory(oppositeDirectionCount, oppositeDirection)
		}
	}
}

func (r *Robot) GetCurrentDirectionString() string {
	switch r.currentDirection {
	case "S":
		return "SOUTH"
	case "E":
		return "EAST"
	case "N":
		return "NORTH"
	case "W":
		return "WEST"
	}
	return "error"
}

// func (r *Robot) GetCurrentDirectionIndex() int {
// 	return r.currentDirectionIndex
// }

func (r *Robot) ChangeDirection() {
	if r.currentDirectionIndex >= 0 && r.currentDirectionIndex < len(r.direction) {
		if r.invert {
			r.currentDirection = r.direction[len(r.direction)-(r.currentDirectionIndex+1)]
		} else {
			r.currentDirection = r.direction[r.currentDirectionIndex]
		}
	} else {
		r.currentDirection = "blocked"
	}
}

func (r *Robot) IncreaseCurrentDirectionIndex() {
	r.currentDirectionIndex++
}

func (r *Robot) ResetCurrentDirectionIndex() {
	r.currentDirectionIndex = 0
}

func (r *Robot) SetCurrentDirection(direction string) {
	r.currentDirection = direction
}

func (r *Robot) GetCurrentDirection() string {
	return r.currentDirection
}

func (r *Robot) GetMovement() []string {
	return r.movement
}

func (r *Robot) GetPosX() int {
	return r.posX
}

func (r *Robot) GetPosY() int {
	return r.posY
}

func (r *Robot) IsInvert() bool {
	return r.invert
}

func (r *Robot) IsDrunk() bool {
	return r.drunk
}

func (r *Robot) DrinkBeer() {
	r.drunk = !r.drunk
}

func (r *Robot) CircuitInverter() {
	r.invert = !r.invert
}

func (r *Robot) IsLooping() bool {
	return r.loop
}

File: blunder-python/robot.py
class Robot:
    def __init__(self, x, y):
        self.x = int(x)
        self.y = int(y)
        self.invert = False
        self.drunk = False
        self.current_direction = "S"
        self.direction = ["S", "E", "N", "W"]
        self.current_direction_index = 0
        self.potential_loop = False
        self.loop = False
        self.memory = []
        self.movement = []

    def move(self, x, y):
        self.x = x
        self.y = y
        self.movement.append(self.get_current_direction_string())
        self.memorize_movement()
    
    def get_opposite_direction(self, direction):
        if direction == "S":
            return "N"
        elif direction == "E":
            return "W"
        elif direction == "N":
            return "S"
        elif direction == "W":
            return "E"

    def memorize_movement(self):
        if len(self.memory) == 0:
            self.memory.append(self.current_direction)
        else:
            if self.memory[0] == self.current_direction: # The robot continues his path
                self.memory.append(self.current_direction)
            elif self.memory[0] == self.get_opposite_direction(self.current_direction): # The robot comes back
                self.memory.append(self.current_direction)
                self.is_stuck_in_loop()
            else: # The robot can't be in loop in this case
                self.memory = []
                self.potential_loop = False # Reset
                self.memory.append(self.current_direction) # It starts his new journey

    def clean_initial_direction_from_memory(self, opposite_direction_count, opposite_direction):
        self.memory = []
        for i in range(opposite_direction_count):
            self.memory.append(opposite_direction)

    def is_stuck_in_loop(self):
        initial_direction = self.memory[0]
        initial_direction_count = 0
        opposite_direction = self.get_opposite_direction(initial_direction)
        opposite_direction_count = 0
        for move in self.memory:
            if move == initial_direction:
                initial_direction_count += 1
            elif move == opposite_direction:
                opposite_direction_count += 1
            else:
                return "Memory is compromised..."
        if initial_direction_count == opposite_direction_count:
            if self.potential_loop:
                self.loop = True
            else:
                self.potential_loop = True
                self.clean_initial_direction_from_memory(opposite_direction_count, opposite_direction)
    
    def get_current_direction_string(self):
        if self.current_direction == "S":
            return "SOUTH"
        elif self.current_direction == "E":
            return "EAST"
        elif self.current_direction == "N":
            return "NORTH"
        elif self.current_direction == "W":
            return "WEST"
    
    def change_direction(self):
        if self.current_direction_index >= 0 and self.current_direction_index < len(self.direction):
            if self.invert:
                self.current_direction = self.direction[len(self.direction)-(self.current_direction_index+1)]
            else:
                self.current_direction = self.direction[self.current_direction_index]
        else:
            self.current_direction = "blocked"
    
    def increase_current_direction_index(self):
        self.current_direction_index += 1

    def reset_current_direction_index(self):
        self.current_direction_index = 0
    
    def set_current_direction(self, direction):
        self.current_direction = direction

    def get_current_direction(self):
        return self.current_direction

    def get_movement(self):
        return self.movement

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def is_invert(self):
        return self.invert

    def is_drunk(self):
        return self.drunk
    
    def drink_beer(self):
        self.drunk = not(self.drunk)

    def circuit_inverter(self):
        self.invert = not(self.invert)

    def is_looping(self):
        return self.loop
